src/ExecutionPathTracer/ExecutionPathTracer.cs

```diff
@@ -97,33 +65,26 @@ public void OnOperationStartHandler(ICallable operation, IApplyData arguments)
         /// </summary>
         public void OnOperationEndHandler(ICallable operation, IApplyData result)
         {
-            this.currentDepth--;
+            if (this.operations.Count <= 1) return;
 
-            // If `compositeTracer` is not null, handle the incoming operation recursively.
-            if (this.compositeTracer != null)
-            {
-                // If the current operation is the composite operation we started with, append
-                // the operations traced out by the `compositeTracer` to the current list of
-                // operations and reset.
-                if (operation == this.currCompositeOp)
-                {
-                    this.AddCompositeOperations();
-                    this.currCompositeOp = null;
-                    this.compositeTracer = null;
-                    return;
-                }
-
-                // Pass operations down to it for handling
-                this.compositeTracer.OnOperationEndHandler(operation, result);
-                return;
-            }
+            var currentOperation = this.operations.Pop();
+            if (currentOperation == null) return;
 
-            // Add latest parsed operation to list of operations, if not null
-            if (this.currentOperation != null)
-            {
-                this.operations.Add(this.currentOperation);
-                this.currentOperation = null;
-            }
+            Operation? parentOp = this.operations.Peek();
```

Comment:
As noted above, it would help to clarify under what conditions the element popped from the stack is expected to be null.